### PR TITLE
Enable TS typechecking in OperatorUI tests

### PR DESCRIPTION
**Pull request by rupurt**
- Use same reducers pattern as explorer
- Add & use @chainlink/ts-test-helpers package (partialAsFull)
- Convert all reducers to TS
- Convert all reducer tests to TS
- Convert all selectors to TS
- Convert all selector tests to TS
- Remove unrequired global definitions
- Use middleware to test async actions
- Add setup:test & setup:dependencies tasks to simply Dockerfile + CI
- Use xlarge when building operator-ui in CI

### CODE REVIEW

File: operator_ui/src/reducers/actions.ts [hunk 1]
```diff
@@ -31,87 +52,293 @@ export interface NotifyErrorAction {
   }
 }
 
+/**
+ * NOTIFY_ERROR_MSG
```
Comment (HenryNguyen5) [inline operator_ui/src/reducers/actions.ts]: How do you feel about grouping action types by enum?

File: operator_ui/src/reducers/notifications.ts [hunk 2]
```diff
@@ -83,6 +76,15 @@ export default function(state: State = initialState, action: Actions) {
   }
 }
 
+export type TextNotification = string
+
+export interface ComponentNotification {
+  component: React.FC<any>
+  props: any
```
Comment (HenryNguyen5) [inline operator_ui/src/reducers/notifications.ts]: Could we get a stricter type here?

File: core/chainlink-sgx.Dockerfile [hunk 3]
```diff
@@ -10,22 +10,12 @@ WORKDIR /chainlink
...
-COPY explorer/client/package.json ./explorer/client/
-COPY explorer/package.json ./explorer/
-COPY operator_ui/package.json ./operator_ui/
-COPY styleguide/package.json ./styleguide/
-COPY tools/json-api-client/package.json ./tools/json-api-client/
-COPY tools/local-storage/package.json ./tools/local-storage/
-COPY tools/redux/package.json ./tools/redux/
-COPY evm/v0.5/package.json ./evm/v0.5/
-COPY evm/package.json ./evm/
+ADD . ./
```
[the remark this answers is not in the document]
Comment (rupurt) [reply core/chainlink-sgx.Dockerfile]: Argg, this is really brittle and annoying. Every time I add a new package I have to add it in a few different places. Is this something Lerna could help with?

File: core/chainlink-sgx.Dockerfile [hunk 4]
```diff
@@ -10,22 +10,12 @@ WORKDIR /chainlink
...
 
+# Install yarn
```
[the remark this answers is not in the document]
Comment (rupurt) [reply core/chainlink-sgx.Dockerfile]: Updated comments to `Install yarn dependencies`

File: operator_ui/__tests__/actions.test.ts [hunk 5]
```diff
@@ -1,16 +1,16 @@
...
 import thunk from 'redux-thunk'
-import isoDate, { MINUTE_MS } from 'test-helpers/isoDate'
-import globPath from 'test-helpers/globPath'
-
-const middlewares = [thunk]
-const mockStore = configureStore(middlewares)
+import * as actions from 'actions'
+import jsonApiJobSpecFactory from '../support/factories/jsonApiJobSpec'
+import jsonApiJobSpecRunFactory from '../support/factories/jsonApiJobSpecRun'
+import isoDate, { MINUTE_MS } from '../support/test-helpers/isoDate'
```
[the remark this answers is not in the document]
Comment (rupurt) [reply operator_ui/__tests__/actions.test.ts]: Fine with the change but it's unrelated to this PR

File: operator_ui/__tests__/actions.test.ts [hunk 6]
```diff
@@ -31,17 +31,29 @@ describe('fetchJob', () => {
 
     global.fetch.getOnce(globPath(`/v2/specs/${jobSpecId}`), jobSpecResponse)
 
+    const testMiddleware: Middleware = () => next => action => {
+      next(action)
+
+      if (action.type === 'UPSERT_JOB') {
```
[the remark this answers is not in the document]
Comment (rupurt) [reply operator_ui/__tests__/actions.test.ts]: Good call

File: core/chainlink-sgx.Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (HenryNguyen5) [reply core/chainlink-sgx.Dockerfile]: Yeah, https://github.com/lerna/lerna/tree/master/commands/list#--toposort

[the remark this answers is not in the document]
Comment (rupurt) [reply core/chainlink-sgx.Dockerfile]: Cool. That's a nice path to improving the situation

File: operator_ui/__tests__/components/Cards/TokenBalance.test.tsx [hunk 7]
```diff
@@ -1,11 +1,11 @@
...
-import { mount } from 'enzyme'
 import React from 'react'
+import TokenBalanceCard from '../../../src/components/Cards/TokenBalance'
+import { mount } from 'enzyme'
 
 describe('components/Cards/TokenBalance', () => {
   it('renders the title and a loading indicator when it is fetching', () => {
     const wrapper = mount(
-      <TokenBalanceCard title="Ether Balance" value={null} />,
+      <TokenBalanceCard title="Ether Balance" value={undefined} />,
```
[the remark this answers is not in the document]
Comment (rupurt) [reply operator_ui/__tests__/components/Cards/TokenBalance.test.tsx]: It's currently marked as optional but I like the explicitness of passing it in. How do you feel about updating the test title instead? Something like

```
'renders the title and a loading indicator value is undefined indicating that it's fetching'
```

File: operator_ui/jest.config.js [hunk 8]
```diff
@@ -1,6 +1,13 @@
 module.exports = {
-  moduleDirectories: ['node_modules', '<rootDir>/src/', '<rootDir>/support/'],
+  preset: 'ts-jest/presets/js-with-ts',
```
[the remark this answers is not in the document]
Comment (rupurt) [reply operator_ui/jest.config.js]: There are component + container tests still in JS

File: operator_ui/jest.config.js [hunk 9]
```diff
@@ -1,6 +1,13 @@
...
+  moduleDirectories: [
+    'node_modules',
+    '<rootDir>/src/',
+    '<rootDir>/support/',
+    '<rootDir>/__tests__',
+  ],
   setupFiles: ['<rootDir>/jest.setup.js'],
+  transformIgnorePatterns: ['node_modules/(?!(@chainlink/json-api-client)/)'],
```
[the remark this answers is not in the document]
Comment (rupurt) [reply operator_ui/jest.config.js]: ✂️

File: operator_ui/src/middleware/explorerConnection.ts [hunk 10]
```diff
@@ -8,20 +8,24 @@ import { Actions } from 'reducers/actions'
...
 ): Middleware {
   const explorerConnectionMiddleware: Middleware = store => next => (
     action: Actions,
   ) => {
     // dispatch original action right away
     next(action)
 
+    // DEV NOTE:
+    // `document` can't be accessed in the NodeJS build environment of the parent closure.
+    const initCookie: string = cookie || (document && document.cookie) || ''
```
[the remark this answers is not in the document]
Comment (rupurt) [reply operator_ui/src/middleware/explorerConnection.ts]: I can't remove those ambient definitions without movie calls to `document` to run time. If it's a default arg it fails at build time.

**Pull request merged**